### PR TITLE
oversampling

### PR DESCRIPTION
**Pull request by generall**

### CODE REVIEW

File: lib/api/src/grpc/conversions.rs [hunk 1]
```diff
@@ -337,6 +337,7 @@ impl From<QuantizationSearchParams> for segment::types::QuantizationSearchParams
             rescore: params
                 .rescore
                 .unwrap_or(default_quantization_rescore_value()),
+            oversampling: None, // TODO: Add `oversampling` to gRPC!
```
Comment (ffuugoo) [inline lib/api/src/grpc/conversions.rs:340]: __TODO__

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 2]
```diff
@@ -228,66 +228,63 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
...
-                .map(|q| q.rescore)
-                .unwrap_or(default_quantization_rescore_value());
-            if quantized && if_rescore {
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        match quantization_params {
+            Some(quantization_params) if quantization_params.rescore => {
+                // TODO: Figure a way to make this calculation without `as` conversions? 🤔
```
Comment (ffuugoo) [inline lib/segment/src/index/hnsw_index/hnsw.rs:267]: __TODO?__ 🤔

**Review by ffuugoo — CHANGES_REQUESTED** [review]
Marking a few TODOs for visibility.

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 3]
```diff
@@ -219,75 +219,80 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
...
-                search_result
-            }
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        if quantized && quantization_params.rescore {
+            let oversampled_top = quantization_params
+                .oversampling
+                .map(|x| (top as f64 * x.max(1.0)) as usize)
```
Comment (ffuugoo) [inline lib/segment/src/index/hnsw_index/hnsw.rs:277]: Should it be `x.min(1.0)`? 🤔

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 4]
```diff
@@ -219,75 +220,80 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
...
-            }
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        if quantized && quantization_params.rescore {
+            let oversampled_top = quantization_params
+                .oversampling
+                .map(|x| (top as f64 * max(x, 1.0)) as usize)
+                .unwrap_or(top);
```
Comment (ffuugoo) [inline lib/segment/src/index/hnsw_index/hnsw.rs:279]: What about this? More verbose, but easier to read, IMO.  🤔

```suggestion
            let oversampling = quantization_params
                .oversampling
                .unwrap_or(1.0);

            // Apply oversampling if necessarry (and ignore oversampling values less than `1.0`)
            let oversampled_top = if oversampling > 1.0 {
                (oversampling * top as f64) as usize
            } else {
                top
            };
```

Comment (ffuugoo) [reply lib/segment/src/index/hnsw_index/hnsw.rs]: > What about this?

I prefer `if` blocks. `if`s are trivial to read, and `bool::then_some` is a form of an IQ test, IMO. 🙈

**Review by ffuugoo — APPROVED** [review]
LGTM.

File: lib/api/src/grpc/conversions.rs (hunk 1 above)
Comment (agourlay) [reply lib/api/src/grpc/conversions.rs:340]: Why is oversampling not propagated in this direction?

File: lib/segment/src/index/hnsw_index/hnsw.rs [hunk 5]
```diff
@@ -219,75 +219,83 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
...
-                re_scored.sort_unstable();
-                re_scored
+        let Some(graph) = &self.graph else {
+            return Vec::new();
+        };
+
+        if quantized && quantization_params.rescore {
+            let oversampling = quantization_params.oversampling.unwrap_or(1.0);
+
+            let oversampled_top = if oversampling > 1.0 {
```
Comment (agourlay) [inline lib/segment/src/index/hnsw_index/hnsw.rs:277]: maybe oversampling < 1.0 should be a client side error

Comment (generall) [reply lib/segment/src/index/hnsw_index/hnsw.rs:277]: Ideally, it should a validation error. But we don't have validation in search yet

Comment (timvisee) [reply lib/segment/src/index/hnsw_index/hnsw.rs:277]: Implemented in <https://github.com/qdrant/qdrant/pull/2100>

**Pull request merged**